gateway/gateway.go

```diff
@@ -586,23 +583,121 @@ func (gateway *HandleT) beaconBatchHandler(w http.ResponseWriter, r *http.Reques
 	gateway.beaconHandler(w, r, "batch")
 }
 
+func (gateway *HandleT) checkAndAddToWebRequestQ(w *http.ResponseWriter, r *http.Request, reqType string, payload []byte, writeKey string) error {
+	var errorMessage = ""
+	if reqType == "import" {
+		errorMessage = gateway.userWebImportHandler(w, r, "batch", payload, writeKey)
+	} else {
+		done := make(chan string, 1)
+		gateway.AddToWebRequestQ(r, w, done, reqType, payload, writeKey)
+		errorMessage = <-done
+	}
+	if errorMessage != "" {
+		return errors.New(errorMessage)
+	}
+	return nil
+}
+
+func (gateway *HandleT) getPayloadAndWriteKey(w http.ResponseWriter, r *http.Request, reqType string) ([]byte, string, error) {
+	var sourceFailStats = make(map[string]int)
+	var err error
+	writeKey, _, ok := r.BasicAuth()
+	if !ok || writeKey == "" {
+		err = errors.New(response.NoWriteKeyInBasicAuth)
+		misc.IncrementMapByKey(sourceFailStats, "noWriteKey", 1)
+		gateway.updateSourceStats(sourceFailStats, "gateway.write_key_failed_requests")
+		return []byte{}, "", err
+	}
+	payload, err := gateway.getPayloadFromRequest(r)
+	if err != nil {
+		sourceName := gateway.getSourceNameForWriteKey(writeKey)
+		sourceTag := misc.GetTagName(writeKey, sourceName)
+		misc.IncrementMapByKey(sourceFailStats, sourceTag, 1)
+		gateway.updateSourceStats(sourceFailStats, "gateway.write_key_failed_requests")
+		return []byte{}, writeKey, err
+	}
+	return payload, writeKey, err
+}
+
 func (gateway *HandleT) webHandler(w http.ResponseWriter, r *http.Request, reqType string) {
 	gateway.logger.LogRequest(r)
 	atomic.AddUint64(&gateway.recvCount, 1)
+	var err error
+	defer func() {
+		if err != nil {
+			gateway.logger.Debug(err.Error())
+			http.Error(w, response.GetStatus(err.Error()), 400)
+		}
+	}()
+	payload, writeKey, err := gateway.getPayloadAndWriteKey(w, r, reqType)
+	if err != nil {
+		return
+	}
+	err = gateway.checkAndAddToWebRequestQ(&w, r, reqType, payload, writeKey)
```

Comment:
Lets rename this function as well. We are not just adding it in the queue. We are processing the request and getting it back as well. Lets call it checkAndProcessWebRequest?